[WIP] Port nuancier to fedora-messaging

sebwoj opened the pull request:
Signed-off-by: Sebastian Wojciechowski <s.wojciechowski89@gmail.com>

sebwoj committed:
Correct updated filed description when election is updated

Signed-off-by: Sebastian Wojciechowski <s.wojciechowski89@gmail.com>

sebwoj committed:
Correct compared values in edit_election.

Before, function edit_election emited message with information about changed "Election year", "Number of votes possible" and "Number of candidates per user" even though those values hasn't changed. The reason was that we compared different types(string with intiger) because the form "add\edit election" saved mentioned values as text instead of integers. This commit corrects this.

Signed-off-by: Sebastian Wojciechowski <s.wojciechowski89@gmail.com>

sebwoj committed:
Port nuancier to fedora-messaging

Signed-off-by: Sebastian Wojciechowski <s.wojciechowski89@gmail.com>

Zlopez commented:
It looks like twisted is no longer supported on python 3.4. I recommend to drop support for python 3.4.

Zlopez reviewed (commented):
As a publisher it should also have it's own scheme.

sebwoj commented:
Current PR is created based on this [commit](https://github.com/fedora-infra/elections/commit/483491f9e956b5c9d59537362644880b069eafe4), and there are no schemes. So what are the requirements for this task?

Zlopez commented:
In this case we could add scheme later. But if you want to know how to do it, look [here](https://github.com/release-monitoring/anitya/tree/master/anitya_schema) and [here](https://github.com/fedora-infra/bodhi/tree/develop/bodhi/messages/schemas).

pypingou commented on nuancier/lib/model.py:
```diff
@@ -153,10 +153,10 @@ def api_repr(self, version):
-                date_end=self.election_date_end,
-                submission_date_start=self.submission_date_start,
-                submission_date_end=self.submission_date_end,
+                date_start=self.election_date_start.strftime('%Y-%m-%d %H:%M:%S') ,
```
You could also use: ``.isoformat()``

pypingou replied:
no isoformat?

sebwoj replied:
Good hint. Actually format "%Y-%m-%d %H:%M:%S" is wrong here, all fields are `date`, not `datetime`.

Commit:
Remove Python 3.4 from Travis

Fedora messaging is using Twisted and this is supported by python 3.5 or newer.

Zlopez commented:
I just merged this PR https://github.com/fedora-infra/nuancier/pull/128, which should fix the Travis error. Could you rebase this to latest master?

Zlopez reviewed (approved):
Everything looks ok

sebwoj commented:
What about schemes? Do you want me to finish implementation? btw I noticed today that tests are missing [fedora_messaging.testing](https://fedora-messaging.readthedocs.io/en/latest/testing.html#module-fedora_messaging.testing)

Zlopez commented:
@sebwoj You can do the schemes later, but if you want to take some time creating them, it will be be good to have them as part of this PR.
You should add the fedora_messaging testing when it makes sense, like sending message when new wallpaper is added.

sebwoj committed:
Add Python 3 to classifiers in setup.py

Signed-off-by: Sebastian Wojciechowski <s.wojciechowski89@gmail.com>

sebwoj commented:
@Zlopez @pypingou On this stage I would like to know your opinion about the change. I know that this PR is not finished.

sebwoj commented:
1 error - left intentionally. Also it looks like this 2 failures are caused by test architecture(tests are expecting particular year, but because 2019 is about to end, there are some dates with 2020).

Zlopez commented:
In the `admin_process_review` it looks like the `mock_sends` is missing and publish has timeout because of this.

The second error in `admin_edit` function looks like the issue you described above.

jontrossbach commented on nuancier/lib/model.py:
```diff
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
 #
-# Copyright © 2013-2017  Red Hat, Inc.
+# Copyright © 2013-2019  Red Hat, Inc. and others.
```
Should this be 2020?

Zlopez commented:
@jontrossbach You are missing the mock function for fedora_messaging. Look at https://fedora-messaging.readthedocs.io/en/stable/testing.html it should guide you in the right direction.

The timeout is happening because it doesn't have any rabbitmq server to connect to, the mock function will prevent this.

jontrossbach reviewed (changes requested):
@sebwoj @Zlopez I did a PR to the fork/branch that this PR is coming from: see [here](https://github.com/sebwoj/nuancier/pull/1). I think the recommended changes might be good to include in this PR.

Zlopez commented:
This look good. Let's wait for @sebwoj to merge these changes.

jontrossbach commented:
Hi all, I've been using these changes quite a bit already with some side projects I've been working on and I've noticed something strange in the way nuancier_schema is set up. For me to be able to deploy this application in OpenShift with these changes I had to adjust how OpenShift sees the path to this module by either explicitly adding it to the $PATH or (my preference) moving the whole module up one directory. What are other peoples thoughts on this?

Maybe there is a good reason for nuancier_schema to be set up the way it is and I am just missing it but I think my suggestion is to move the whole module up one directory. 

Also, was nuancier_schema meant to be a git submodule? It almost seems like it was.

bkmgit commented:
Failed builds seem to come from python 2.7, 3.5 and 3.6. Are the changes merged into https://github.com/fedora-infra/nuancier/pull/133 ok? Will there be any more development on this fork?

ryanlerch commented:
Most projects that are currently being ported to fedora-messaging schemas are using a separate repo and python package for the schemas. This allows two things, consumers of the messages can install the schemas without having to install the whole application, and it having it in a separate repo allows it to have a different release cadence than the main project.

There is also a python cookie-cutter recipe that allows quick creation of a schemas repo, and python package,  as well as tests:

https://github.com/fedora-infra/cookiecutter-message-schemas

Zlopez commented:
@ryanlerch I think it would be nice to add this to official fedora messaging documentation.

ryanlerch commented:
> @ryanlerch I think it would be nice to add this to official fedora messaging documentation.

https://fedora-messaging.readthedocs.io/en/stable/messages.html#packaging

ryanlerch commented:
Just FYI too, before i knew this PR existed, i created this repo here:

https://github.com/fedora-infra/nuancier-messages

if it is used, woo! if not, i will get it deleted.

Zlopez commented:
It's better to use separate repo for schemas, it is easier to maintain.

bkmgit commented:
@ryanlerch Thanks for the information.

sebwoj commented:
@pypingou I'm working in Openshift team right now, so I'm out of the context. But feel free to modify/close/update this PR.
